cpp/src/arrow/compute/kernel.cc

```diff
@@ -430,6 +431,25 @@ bool InputType::Matches(const Datum& value) const {
   return Matches(*value.type());
 }
 
+bool InputType::MatchesDeviceAllocationType(const Datum& value) const {
+  DCHECK(Matches(value));
```

Comment:
I will probably remove this `DCHECK` and keep the ones in `KernelSignature::MatchesDeviceAllocationTypes`.